### ISSUE TITLE
State change consideration in methods.

### ISSUE THREAD

**Issue by resistdesign** [gajus/youtube-player#21]
@gajus It seems that the API methods are promisified but only consider ready state and not state changes, like `playVideo()` should wait until the player state is `1` before resolving.

This make this library very hard to use.

Or am I missing something?

**Comment by gajus** [comment]
I haven't touched this code for a long time, therefore I'd need to read through the YouTube player docs to understand the implications. Given that you work with this library at the moment, I think you are the best to judge what is a desired behaviour and make the respective improvements.

**Comment by resistdesign** [comment]
I have a need to use it but it needs to work better.

I'm considering a rewrite with modern ES7 features like `async/await` and that sort of thing while considering the various player statuses that makes sense to wait for based on the operation.

How do you feel about a rewrite?

**Comment by resistdesign** [comment]
I'm honestly thinking less dynamic and more hardcoded, meaning choosing certain features and supporting them with actual class methods. Then maybe dynamically adding the rest of the methods as a means of access but not making any guarantees as to how they might function. Also, the use of `on` from `sister` is ok, but it doesn't seem like `off` is implemented and I can see a need to remove listeners as well. That may be just a matter of exposing the `add/removeEventListener` methods in a smart way or just encouraging direct access to the underlying YT Player instance for certain scenarios.

**Comment by resistdesign** [comment]
Additionally, I think the exported class would need a way of injecting the YT Player instance SO THAT it can be stubbed for testing purposes. This way a unit test can more easily be written that doesn't require network access or even an actual browser. If you know better about the current implementation, where it applying to testing without the actual YT Player instance, please let me know.

**Comment by gajus** [comment]
> I'm considering a rewrite with modern ES7 features like async/await and that sort of thing while considering the various player statuses that makes sense to wait for based on the operation.
>
> How do you feel about a rewrite?

Go for it. I am happy to add you as a collaborator to the project.

**Comment by gajus** [comment]
It would be nice to discuss an outline of what are the goals of the rewrite and what benefit they offer over the existing implementation.

**Comment by resistdesign** [comment]
I agree, here are the goals with benefits as I can see right now:

1. A guarantee for a select few methods that their returned promise won't resolve until the player is in an expected state so that state inspection isn't a required, extra step during implementation.
2. Improved testability through the ability to stub the YT Player instance that methods use internally.
3. Modern code using `async/await` and spread operators or other modern syntax as needed so that the code is up to date and taking advantage of modern best practices.
4. Object oriented architecture for a more clearly defined and understood structure.
5. An asynchronous, public accessor (getter) for the underlying YT Player instance to provide lower level access for operations like adding and removing event listeners without the need to wrap or inhibit such functionality.

Let me know what you think of this list. Do you have additional things in mind? Do you agree with these items?

**Comment by gajus** [comment]
> A guarantee for a select few methods that their returned promise won't resolve until the player is in an expected state so that state inspection isn't a required, extra step during implementation.

Sounds good. Do you have specific examples?

You have mentioned that `playVideo` should not resolve until player state becomes `1`. Anything else?

On the subject of `playVideo`, how is buffering and other intermittent states going to be handled?

Video can go into play state and instantly into buffering. In such case, it makes little difference whether `playVideo` resolves instantly or after player turns into PLAYING state.

> Improved testability through the ability to stub the YT Player instance that methods use internally.

:+1:

> Modern code using async/await and spread operators or other modern syntax as needed so that the code is up to date and taking advantage of modern best practices.

Please make sure to pull the latest version of `eslint-config-canonical`.

> Object oriented architecture (using classes) for a more clearly defined and understood structure.
Including doc comments for public members.

For public facing API, yes, sure. Though the logic should be implemented using functional style, as it is now.

> An asynchronous, public accessor (getter) for the underlying YT Player instance to provide lower level access for operations like adding and removing event listeners without the need to wrap or inhibit such functionality.

I would rather not do this and focus on providing an API for those uses cases. YT API surface is not big.

**Comment by resistdesign** [comment]
> Sounds good. Do you have specific examples?

> You have mentioned that playVideo should not resolve until player state becomes 1. Anything else?

> On the subject of playVideo, how is buffering and other intermittent states going to be handled?

> Video can go into play state and instantly into buffering. In such case, it makes little difference whether playVideo resolves instantly or after player turns into PLAYING state.

I'm looking to ensure that pause is in the pause state, seekTo results in either play or pause based on how it started and to ensure that play results in a play state. Maybe even cue resolving once the player is in the play state.

As far as play jumping to buffering, I think I might implement an interval to check to see that the play state has "stabilized" with a timeout that will reject. Maybe this functionality could be in some optional precision mode or something.

Still brainstorming on specific solutions there and I will have to play with the exact behavior.

> Please make sure to pull the latest version of eslint-config-canonical.

Will do!

> For public facing API, yes, sure. Though the logic should be implemented using functional style, as it is now.

I'll be looking to provide functional code where it makes sense and OOP classes where they make sense such public facing code as you suggest.

> I would rather not do this and focus on providing an API for those uses cases. YT API surface is not big.

Ok, I think it will suffice to expose add AND remove event listener features.

Let me know what you think and then I'd like to move forward because I actually have use for this right now :)

Make me a collaborator if you would like, I think it might be easier to deal with direct branches rather than "forking around" :)

**Comment by gajus** [comment]
> Make me a collaborator if you would like, I think it might be easier to deal with direct branches rather than "forking around" :)

Done

**Comment by resistdesign** [comment]
@gajus Ok so my plan is to have a branch for this issue called `fix/21` (or something like that). And then have sub-task branches, with just a few code changes at a time, that I will PR into the issue branch for your review. Once we have everything squared away and stable we can go to `master` and release a new version.

Does that work for you? Any thoughts???

**Comment by gajus** [comment]
Sounds like a plan. :-)

**Comment by resistdesign** [comment]
@gajus I have an `eslint` rule issue:

```bash
youtube-player/test/functionNames.js
  4:40  warning  Identifier name 't' is too short (< 2)  id-length
```

Do you want to disable the rule or change the convention of using `t` in tests?

**Comment by resistdesign** [comment]
@gajus I created a slack team for easy communication: https://youtubeplayerapi.slack.com

You should have an email invite.

**Comment by resistdesign** [comment]
@gajus Let me know what you want me to do with that ESLint issue from above when you get a chance.

**Comment by gajus** [comment]
> Do you want to disable the rule or change the convention of using t in tests?

Disable the rule for the tests.

> resistdesign commented 3 days ago

Sorry, if I don't reply within 24 hours, chances are that your message got lost between notifications.

**Comment by resistdesign** [comment]
@gajus Np, ok, I will disable the rule :)

**Comment by resistdesign** [comment]
@gajus What's you're preferred approach to stepping outside of the list of functions for functions that will wait for a particular state before resolving?

**Comment by gajus** [comment]
> @gajus What's you're preferred approach to stepping outside of the list of functions for functions that will wait for a particular state before resolving?

Please rephrase the question. I don't understand what you are referring to.

**Comment by resistdesign** [comment]
Currently, there is an array of function names that is iterated through in order to create a promise version of all function for the YT Player API. However, for a few of those functions, the plan is to add the additional constraint that the player must be in a specified state related to the function called before the promise will be resolved.

So the question is, if I am to apply this additional functionality for only a few functions, then what are your suggestions for treating those functions differently.

Basically, I just want to make sure that you aren't aware of any possible harm/caveates that this could cause or if you have any preferences around this type of augmentation.

My initial thought is to simply use a map from function names to arrays of acceptable states and a condition that operates normally if then iterated function name is not found in the map.

**Comment by gajus** [comment]
> My initial thought is to simply use a map from function names to arrays of acceptable states and a condition that operates normally if the iterated function name is not found in the map.

What you are describing sounds like a state machine, e.g. https://github.com/jakesgordon/javascript-state-machine

Should work.
